### PR TITLE
Added timezone selection on the agenda page

### PR DESCRIPTION
**Pull request by MomirMilutinovic**
Users can now select their timezone, and all times on the agenda will be shifted according to their selection.

- The talks query can now take the timezone into account
- Added a property to the AgendaTalksList component that sets the timezone offset

Note: Times can roll over into sometime after the conference. This prevents them from being found on the agenda page since the user can only see the preset conference days.

### CODE REVIEW

File: src/server/apollo/resolvers/Query.ts [hunk 1]
```diff
@@ -2,13 +2,31 @@ import { assertConferenceDay } from "../../../const";
...
+import { addMinutes } from "../../../utils/addMinutes";
 
 export const Query: QueryResolvers = {
   talks: async (_root, args, { dataSources }) => {
     const talks = await dataSources.pretalx.getAllTalks();
 
+    let newTalks = [];
+
+    // Copy all the talks
+    for (var i = 0; i < talks.length; i++) {
```
Comment (travigd) [inline src/server/apollo/resolvers/Query.ts:14]: I'd prefer
```ts
const newTalks = talks.map((talk) => ({ ...talk }));
```

Nowadays, imperative array manipulation (e.g., using `.push`) has fallen out of style.

File: src/utils/addMinutes.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { isNonNull } from "./null";
+
+// Adds some number of minutes to a time passed as a string in ISO format
+// Null values are treated as zeroes
+export function addMinutes(time: string, minutes: number | null | undefined) {
```
Comment (travigd) [inline src/utils/addMinutes.ts:5]: I think there's already a function to do this in the `date-fns` library. I think that's installed.

**Comment by MomirMilutinovic** [comment]
I have moved the timezone functionality to the front-end. 
The most significant change is that when the time of a talk rolls over to the next day, it is displayed on the page of the previous day (and not the day it technically is on). There is text next to the time that indicates that a rollover happened (e.g. 01:00 The next day).

File: src/components/agenda/AgendaTalksListItem.tsx [hunk 3]
```diff
@@ -30,6 +33,31 @@ export const AgendaTalksListItem = ({
 
   const { title, abstract, startTime, endTime, speakers } = data.talk;
 
+  let startTimeDate = new Date(startTime);
```
Comment (travigd) [inline src/components/agenda/AgendaTalksListItem.tsx:36]: All of this should probably be its own React component.

File: src/components/agenda/AgendaTalksListItem.tsx [hunk 4]
```diff
@@ -16,9 +17,11 @@ import {
 export const AgendaTalksListItem = ({
   talkId,
   noTopBorder,
+  zoneOffset,
 }: {
   talkId: string;
   noTopBorder?: boolean;
+  zoneOffset: number;
```
Comment (travigd) [inline src/components/agenda/AgendaTalksListItem.tsx:24]: nit: should probably be `timeZoneOffset`

File: src/pages/agenda/[day].tsx [hunk 5]
```diff
@@ -37,6 +41,10 @@ const Agenda: NextPage = () => {
     return <Error statusCode={404} />;
   }
 
+  function onChange(option: any) {
```
Comment (travigd) [inline src/pages/agenda/[day].tsx:44]: Typically using `any` is a ~~bad thing~~ but the typing here would be annoying so this is fine.

File: src/pages/agenda/[day].tsx [hunk 6]
```diff
@@ -47,7 +55,18 @@ const Agenda: NextPage = () => {
...
+      <Select
+        options={timezoneOptions}
+        onChange={onChange}
+        value={zoneOffsetOption}
+        placeholder="Choose a timezone"
+      />
+      <VSpace />
+      <AgendaTalksList
+        conferenceDay={day}
+        zoneOffset={zoneOffsetOption.value}
```
Comment (travigd) [inline src/pages/agenda/[day].tsx:68]: It would actually be better to pass this around using React Context.

File: src/components/date/Time.tsx [hunk 7]
```diff
@@ -12,5 +12,5 @@ import { format, parseISO } from "date-fns";
  */
 export const Time = ({ time }: { time: Date | string }) => {
   const date = typeof time === "string" ? parseISO(time) : time;
-  return <>{format(date, "HH:mm")} UTC</>;
+  return <>{format(date, "HH:mm")}</>;
```
Comment (travigd) [inline src/components/date/Time.tsx:15]: I think this should still display the current timezone. This is easier to do with React Context (see below).

**Comment by MomirMilutinovic** [comment]
I have reimplemented the timezone functionality with React context as @travigd suggested. The timezone is set in the dropdown on the agenda page.

**Comment by travigd** [comment]
Looks good to me. Next steps are storing the users time zone selection in a cookie or `localStorage`. Also it would be nice to have better labels for the timezones (e.g. `UTC-7:00 Pacific Daylight Time`). Things get messy with daylight savings in the states but since we're only doing the even for a week we don't really need to worry about that (except for labels). Finally, the timezone selection is a bit ugly and doesn't need to be so big and front-and-center, but we're definitely trying to get the functionality there before we make it look pretty.

As always, than you so much for all your work on this! It's presentable now so I'm going to merge it. If you have the time please feel free to address those followup this.

**Pull request merged**